atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStore.java

```diff
@@ -160,20 +162,223 @@ public static long getMillisForTimestamp(KeyValueService kvs, long timestamp) {
+     *   2) the result has an associated ts greater than timestamp; there is no need to range scan after result.millis()
+     *   (note that this is generally less than testValue)
+     *   3) otherwise; there is no need to range scan before testValue, but if there is no valid entry in the range
+     *   then betweenBounds.millis() is the best approximation for the real time in milliseconds corresponding to
+     *   timestamp.
+     *
+     * Once both the upper bound and the lower bound have an associated timestamp, we determine the next test value
+     * by assuming a linear correlation between time in milliseconds and timestamps, and interpolate the next
+     * candidate in the standard way: y = d(x)/d(y) (x - x_1) + y_1. Otherwise, we simply take the midpoint.
+     *
+     * Note  that, as an optimisation, the initial test value is expected to be equal to lower bound in this
+     * implementation, but in cases lower bound is extremely low, the test value will be relatively close to the upper
+     * bound.
+     *
+     * @return bounds for range scan if range scan should be performed, and a value in milliseconds that is the best
+     * approximation if the range scan returns no results.
      */
-    public static long getMillisForTimestampIfNotPunchedBefore(KeyValueService kvs, long timestamp, long lowerBound) {
-        if (get(kvs, Math.max(0L, lowerBound)) < timestamp) {
-            return getMillisForTimestamp(kvs, timestamp);
-        } else {
-            return lowerBound;
+    private static BoundsOrMillis getBoundsForRangeScan(
+            KeyValueService kvs, long timestamp, Bounds initialBounds, Optional<Long> lowerBoundTimestamp) {
+        MillisAndMaybeTimestamp upperBound = ImmutableMillisAndMaybeTimestamp.builder()
+                .millis(initialBounds.upper())
+                .build();
+        MillisAndMaybeTimestamp lowerBound = ImmutableMillisAndMaybeTimestamp.builder()
+                .millis(initialBounds.lower())
+                .timestamp(lowerBoundTimestamp)
+                .build();
+        long testValue = Math.max(initialBounds.lower(), initialBounds.upper() - 7 * MAX_RANGE_SCAN_SIZE);
+        Optional<MillisAndMaybeTimestamp> millisLowEstimate = Optional.empty();
+        while (upperBound.millis() - lowerBound.millis() > MAX_RANGE_SCAN_SIZE) {
+            MillisAndMaybeTimestamp betweenBounds = getOlder(kvs, testValue);
+
+            if (betweenBounds.timestampSatisfies(ts -> ts == timestamp)) {
+                return noRangeScan(betweenBounds);
+            } else if (betweenBounds.timestampSatisfies(ts -> ts > timestamp)) {
+                upperBound = betweenBounds;
+            } else {
+                millisLowEstimate = Optional.of(betweenBounds);
+                lowerBound = ImmutableMillisAndMaybeTimestamp.builder()
+                        .from(betweenBounds)
```

Comment:
This doesn't really save us any code 😅  just get the timestamp from betweenBounds?